src/multiplex/multiplex.rs

```diff
@@ -377,75 +451,175 @@ impl<S, T, E> Multiplex<S, T>
+            try!(self.dispatch.transport().write(frame));
+            self.blocked_on_flush.wrote_frame();
+
+            e.remove();
+        } else {
+            trace!("exchange does not exist; id={:?}", id);
         }
 
         Ok(())
     }
 
     fn write_in_body(&mut self) -> io::Result<()> {
-        trace!("write in body chunks; handlers={}", self.in_bodies.len());
+        trace!("write in body chunks");
 
         self.scratch.clear();
 
         // Now, write the ready streams
         'outer:
-        for (id, body) in &mut self.in_bodies {
+        for (&id, exchange) in &mut self.exchanges {
             trace!("   --> checking request {:?}", id);
-            while self.transport.poll_write().is_ready() {
-                match body.poll() {
+
+            loop {
+                if !self.dispatch.transport().poll_write().is_ready() {
+                    self.blocked_on_flush.transport_not_write_ready();
+                    break 'outer;
+                }
+
+                match exchange.try_poll_in_body() {
                     Ok(Async::Ready(Some(chunk))) => {
                         trace!("   --> got chunk");
 
-                        // TODO: handle return value
-                        let frame = Frame::Body(*id, Some(chunk));
-                        try!(self.transport.write(frame));
+                        let frame = Frame::Body { id: id, chunk: Some(chunk) };
+                        try!(self.dispatch.transport().write(frame));
+                        self.blocked_on_flush.wrote_frame();
                     }
                     Ok(Async::Ready(None)) => {
                         trace!("   --> end of stream");
 
-                        let frame = Frame::Body(*id, None);
-                        try!(self.transport.write(frame));
-                        // In body stream fully written, remove the entry and
-                        // continue to the next body
-                        self.scratch.push(*id);
-                        continue 'outer;
+                        let frame = Frame::Body { id: id, chunk: None };
+                        try!(self.dispatch.transport().write(frame));
+                        self.blocked_on_flush.wrote_frame();
+
+                        // in_body is fully written.
+                        exchange.in_body = None;
+                        break;
                     }
-                    Err(_) => {
-                        unimplemented!();
+                    Err(error) => {
+                        // Write the error frame
+                        let frame = Frame::Error { id: id, error: error };
+                        try!(self.dispatch.transport().write(frame));
+                        self.blocked_on_flush.wrote_frame();
+
+                        exchange.responded = true;
+                        exchange.in_body = None;
+                        exchange.out_body = None;
+                        exchange.out_deque.clear();
```

Comment:
This appears at the end of `send_out_chunk` and `try_poll_in_body` below I think, perhaps these could be two methods of `self.in_finished()` and `self.out_finished()`? (or something like that)